[DEV-753] Merge feat/config-env-vars into develop

maartends opened the pull request:
This merge implements the feature request for environment variable expansion as per ticket [https://meemoo.atlassian.net/browse/DEV-753](https://meemoo.atlassian.net/browse/DEV-753).

**Note!** This introduces backward incompatible changes to the `ConfigParser`'s API, notably the following attributes:
- `self.chassis_cfg`
- `self.app_cfg`

maartends committed:
Add shebang and coding

(No code-changes yet.)

maartends committed:
[DEV-753] Add minimal testing for configuration

Add tests for our ConfigParser class as to keep everything working.
Run these tests with:
$ pytest -v tests/test_configuration.py

(No code-changes yet.)

maartends committed:
[DEV-753] Make the default a default fn-argument

- The default config file is now a default value to the `config_file`-argument.
- Prefer opening the file and catching the IOError over testing for the file
  (EAFP over LBYL).
- Add more tests and two test config file as resources.

maartends committed:
[DEV-753] Fix tests and reflow configparser

maartends committed:
[DEV-753] Add application cfg + refactor parser

- Make the ConfigParser not forget about application specific configuration.
  For now, this is stored in it's `cfg`-attr. This will change.
- Refactor `__init__` into proper methods.
- Add tests and a resource.

maartends committed:
[DEV-753] Add first working version of env vars

- Plus, add tests and a new resource.

maartends committed:
[DEV-753] Change custom constructor to class

- Change the custom constructor for the `!ENV` tag to a class instead of a function.
- This class now raises a KeyError if a referenced environment variable is
  missing! This is the desired behaviour as chances are very slim that an
  application could be running flawlessly when it is missing some configuration
  options.
- Add a test for missing env vars (although this test doesn't work as of yet).
- Update README to reflect these changes (as our API has changed as well)
- README: also reflowed some sections and changed the line endings to Unix.

RudolfDG committed:
Fix test with missing env vars.
Environment variables got set by previous tests so they weren't missing.
Test now makes use of the monkeypatch fixture to handle env vars.

maartends committed:
[DEV-753] Update ___repr__ of EnvVar

maartends committed:
Replace `config.config` with `config.chassis_cfg`

Makes the distinction between Chassis.py-config and application specific
config more obvious.
Adpapt tests accordingly.
-> Breaking change!

maartends committed:
Replace `config.config` with `config.chassis_cfg`

- commit b303eee54930174ede0c77cb2200fe1eea0dc66d introduced
  backwards incompatible changes. Compensating.

ghost commented:
## DeepCode failed to analyze this pull request
Something went wrong despite trying multiple times, sorry about that.
Please comment this pull request with "Retry DeepCode" to manually retry, or [contact us](https://www.deepcode.ai/feedback?select=4&utm_source=gh_review) so that a human can look into the issue.

maartends committed:
Replace `config.config` with `config.chassis_cfg`

- commit b303eee54930174ede0c77cb2200fe1eea0dc66d introduced
  backwards incompatible changes. Compensating in logging tests.

RudolfDG committed:
Change loglevel to `WARNING` in docs and example.

RudolfDG commented:
Looks good to me, gave it a test drive and it seems to be working nicely. :)

spacid reviewed (changes requested):
- The test class `TestConfigWithMissingEnvVars` will fail if you have set the `LOG_PATH` environment variable manually in the same environment where you run the tests. In order to be more deterministic, it can be temporarily be removed: `monkeypatch.delenv("LOG_PATH", raising=False) `.
- I had a `config.yml` file in the working directory which made class `TestConfigNoFile` fail, as stated in the code. A possibility could be to change the test. Instead of testing the logic if the default config file would not exist, to test the logic if it can't find a certain config file e.g. ConfigParser("nowaythisfileexist.yml").
- To expand upon the [commit of Rudolf](https://github.com/viaacode/chassis.py/commit/f45d59993763daf944983697d9b8ff69d7faf467), the files in `./tests/resources` also have logging level `WARN`. Although these are not being used by the logging tests, and thus no failing tests, it might be advised to adapt those, as these files can also be seen as examples.
- A couple of documentation typos. I will fix them in a commit.
- In `./tests/test_configuration`, `os` is imported but unused. There might be others, perhaps check with a linter.

spacid committed:
[DEV-753] Fix a couple documentation typos

maartends committed:
[DEV-753] Change WARN to WARNING in resources

Reference: https://github.com/viaacode/chassis.py/pull/3#pullrequestreview-393658907

maartends committed:
[DEV-753] Fix config-test as per remarks

Reference: https://github.com/viaacode/chassis.py/pull/3#pullrequestreview-393658907

maartends commented:
In reply to: https://github.com/viaacode/chassis.py/pull/3#pullrequestreview-393658907.

- `LOG_PATH` environment variable: correct and a very good enhancement. Done.
- `TestConfigNoFile`: This was changed to make the tests more deterministic, however -> see comment below.
- `WARN` -> `WARNING`: Done.
- Docs: thanks.
- `import os` removed.

About `TestConfigNoFile`: changing the test-code to providing the ConfigParser with the path to a non-existant file makes the tests more deterministic indeed. However, we are now testing a different case than before: now we provide the ConfigParser with the path to a file that is not there. Calling the ConfigParser with an explicit filepath argument would only ever happen from within the application code, and thus, should raise an exception instead of falling back to a default configuration. I will open an issue to address this in the future (-> https://meemoo.atlassian.net/browse/DEV-790).

(Moreover, the test now relies on the non-existant file to... not exist, however unlikely. :-) )
